### PR TITLE
Fix d2i_ECPKParameters_fp and i2d_ECPKParameters_fp macroes

### LINKED ISSUE #12443: d2i_ECPKParameters_fp throws lots of warnings
I think some of [these macros](https://github.com/openssl/openssl/blob/ce3080e931d77fda3bb4f2d923fcc6cec967d1a3/include/openssl/ec.h#L836) are stale. Below is master but I get the same warnings in 1.1.1, although the code seems to runs fine.

MWE:
```C
#include <openssl/ec.h>
#include <assert.h>

int main(int argc, char **argv) {
    EC_GROUP *group = NULL;
    FILE *fp = NULL;

    assert((fp = fopen("params.der", "rb")) != NULL);
    assert((group = d2i_ECPKParameters_fp(fp, NULL)) != NULL);
    fclose(fp);
    EC_GROUP_free(group);

    return 0;
}
```

```
bbrumley@arizona:/tmp$ rm foo; cc -o foo foo.c -lcrypto -L$HOME/.local/ssl/lib -I$HOME/.local/ssl/include
In file included from foo.c:2:0:
foo.c: In function ‘main’:
/home/bbrumley/.local/ssl/include/openssl/ec.h:841:35: warning: passing argument 2 of ‘ASN1_d2i_fp’ from incompatible pointer type [-Wincompatible-pointer-types]
     (EC_GROUP *)ASN1_d2i_fp(NULL, (char *(*)())d2i_ECPKParameters, (fp), \
                                   ^
foo.c:9:21: note: in expansion of macro ‘d2i_ECPKParameters_fp’
     assert((group = d2i_ECPKParameters_fp(fp, NULL)) != NULL);
                     ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/bbrumley/.local/ssl/include/openssl/ec.h:23:0,
                 from foo.c:1:
/home/bbrumley/.local/ssl/include/openssl/asn1.h:689:7: note: expected ‘void * (*)(void **, const unsigned char **, long int)’ but argument is of type ‘char * (*)()’
 void *ASN1_d2i_fp(void *(*xnew) (void), d2i_of_void *d2i, FILE *in, void **x);
       ^~~~~~~~~~~
In file included from foo.c:2:0:
/home/bbrumley/.local/ssl/include/openssl/ec.h:842:29: warning: passing argument 4 of ‘ASN1_d2i_fp’ from incompatible pointer type [-Wincompatible-pointer-types]
                             (unsigned char **)(x))
                             ^
foo.c:9:21: note: in expansion of macro ‘d2i_ECPKParameters_fp’
     assert((group = d2i_ECPKParameters_fp(fp, NULL)) != NULL);
                     ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/bbrumley/.local/ssl/include/openssl/ec.h:23:0,
                 from foo.c:1:
/home/bbrumley/.local/ssl/include/openssl/asn1.h:689:7: note: expected ‘void **’ but argument is of type ‘unsigned char **’
 void *ASN1_d2i_fp(void *(*xnew) (void), d2i_of_void *d2i, FILE *in, void **x);
       ^~~~~~~~~~~
In file included from foo.c:2:0:
/home/bbrumley/.local/ssl/include/openssl/ec.h:841:35: warning: passing argument 2 of ‘ASN1_d2i_fp’ from incompatible pointer type [-Wincompatible-pointer-types]
     (EC_GROUP *)ASN1_d2i_fp(NULL, (char *(*)())d2i_ECPKParameters, (fp), \
                                   ^
foo.c:9:21: note: in expansion of macro ‘d2i_ECPKParameters_fp’
     assert((group = d2i_ECPKParameters_fp(fp, NULL)) != NULL);
                     ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/bbrumley/.local/ssl/include/openssl/ec.h:23:0,
                 from foo.c:1:
/home/bbrumley/.local/ssl/include/openssl/asn1.h:689:7: note: expected ‘void * (*)(void **, const unsigned char **, long int)’ but argument is of type ‘char * (*)()’
 void *ASN1_d2i_fp(void *(*xnew) (void), d2i_of_void *d2i, FILE *in, void **x);
       ^~~~~~~~~~~
In file included from foo.c:2:0:
/home/bbrumley/.local/ssl/include/openssl/ec.h:842:29: warning: passing argument 4 of ‘ASN1_d2i_fp’ from incompatible pointer type [-Wincompatible-pointer-types]
                             (unsigned char **)(x))
                             ^
foo.c:9:21: note: in expansion of macro ‘d2i_ECPKParameters_fp’
     assert((group = d2i_ECPKParameters_fp(fp, NULL)) != NULL);
                     ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/bbrumley/.local/ssl/include/openssl/ec.h:23:0,
                 from foo.c:1:
/home/bbrumley/.local/ssl/include/openssl/asn1.h:689:7: note: expected ‘void **’ but argument is of type ‘unsigned char **’
 void *ASN1_d2i_fp(void *(*xnew) (void), d2i_of_void *d2i, FILE *in, void **x);
       ^~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by romen**
This PR is the `master` counterpart of #12457, and fixes #12443.

This is the original message for the 1.1.1 commit:

> ## Fix `d2i_ECPKParameters_fp` and `i2d_ECPKParameters_fp` macroes
>
> These functions are part of the public API but we don't have tests
> covering their usage.
> They are actually implemented as macroes and the absence of tests has
> caused them to fall out-of-sync with the latest changes to ASN1 related
> functions and cause compilation warnings.
>
> This commit fixes the public headers to reflect these changes.

In `master` we inherited the defect in the public header.

The test for this uncovered a second bug, dealing with curves without associated OIDs.
This PR (as well as #12457) also includes a fix for this second issue.

## Tasks

- [x] Cherry-pick test from #12457
- [x] Evaluate if #12443 still affects `master`
  - [x] Fix #12433 in `master` if needed
- [x] Pass `*_fp()` tests in Windows

### CODE REVIEW

**Comment by romen** [comment]
The fix touches public API, so mandatory OTC decision label added.

File: crypto/ec/ec_asn1.c [hunk 1]
```diff
@@ -485,7 +485,7 @@ ECPKPARAMETERS *EC_GROUP_get_ecpkparameters(const EC_GROUP *group,
             ECPARAMETERS_free(ret->value.parameters);
     }
 
-    if (EC_GROUP_get_asn1_flag(group)) {
+    if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE) {
```
Comment (paulidale) [inline crypto/ec/ec_asn1.c:488]: Shouldn't this be `&` not `==` here and throughout?

Existing code uses both forms: `==` once and `&` twice.

Comment (romen) [reply crypto/ec/ec_asn1.c:488]: Uhm, this ASN1 flag is actually a single bit, so there shouldn't be a difference. I can go either way. In 1.1.1 this flag is never associated with boolean operators.

Comment (romen) [reply crypto/ec/ec_asn1.c:488]: I actually prefer `==`, it allows symmetry between

~~~C
if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE)
~~~

and 

~~~C
if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_EXPLICIT_CURVE)
~~~

We cannot use boolean operators on `OPENSSL_EC_EXPLICIT_CURVE` as its value is `0x0`

Comment (romen) [reply crypto/ec/ec_asn1.c:488]: Also, we spread around code like

~~~C
EC_GROUP_set_asn1_flag(ret_group, OPENSSL_EC_EXPLICIT_CURVE)
~~~

so I believe there is little point in overengineering for a future where this flag field will ever include additional bits: changing it in the future will break a lot of code.

Comment (romen) [reply crypto/ec/ec_asn1.c:488]: This patch would improve consistency. I dislike the first hunk.

~~~diff
diff --git a/crypto/ec/ec_backend.c b/crypto/ec/ec_backend.c
index 381da71f33..4ee093f06b 100644
--- a/crypto/ec/ec_backend.c
+++ b/crypto/ec/ec_backend.c
@@ -308,7 +308,8 @@ int ossl_ec_group_todata(const EC_GROUP *group, OSSL_PARAM_BLD *tmpl,
         ERR_raise(ERR_LIB_EC, EC_R_INVALID_FORM);
         return 0;
     }
-    encoding_flag = EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE;
+    encoding_flag = EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE ?
+        OPENSSL_EC_NAMED_CURVE : OPENSSL_EC_EXPLICIT_CURVE;
     encoding_name = ec_param_encoding_id2name(encoding_flag);
     if (encoding_name == NULL
         || !ossl_param_build_set_utf8_string(tmpl, params,
diff --git a/providers/implementations/encode_decode/encode_key2any.c b/providers/implementations/encode_decode/encode_key2any.c
index 855866cbfe..1ade5800bb 100644
--- a/providers/implementations/encode_decode/encode_key2any.c
+++ b/providers/implementations/encode_decode/encode_key2any.c
@@ -679,7 +679,7 @@ static int prepare_ec_params(const void *eckey, int nid, int save,
     }
 
     if (curve_nid != NID_undef
-        && (EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE)) {
+        && (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE)) {
         /* The CHOICE came to namedCurve */
         if (OBJ_length(params) == 0) {
             /* Some curves might not have an associated OID */
diff --git a/providers/implementations/encode_decode/encode_key2text.c b/providers/implementations/encode_decode/encode_key2text.c
index f8f9712e96..60f3e13a40 100644
--- a/providers/implementations/encode_decode/encode_key2text.c
+++ b/providers/implementations/encode_decode/encode_key2text.c
@@ -460,7 +460,7 @@ err:
 static int ec_param_to_text(BIO *out, const EC_GROUP *group,
                             OSSL_LIB_CTX *libctx)
 {
-    if (EC_GROUP_get_asn1_flag(group) & OPENSSL_EC_NAMED_CURVE) {
+    if (EC_GROUP_get_asn1_flag(group) == OPENSSL_EC_NAMED_CURVE) {
         const char *curve_name;
         int curve_nid = EC_GROUP_get_curve_name(group);
 
~~~

Comment (paulidale) [reply crypto/ec/ec_asn1.c:488]: If we're just using it for this one flag, the function/field names are wrong.  _named_ would make more sense.

Comment (romen) [reply crypto/ec/ec_asn1.c:488]: I agree the name is not particularly friendly, especially given that other similar functions are named as `*_get_flags()` and handled with bitwise logic. "asn1_flag" is signular, but indeed not very descriptive.

It seems unfortunately it is too late to change this now. Looking back at our history, it seems quite a deliberate design the fact that this is a 1-bit flag.

Before 254ef80db19a6f6610ef07535efd699cef7aa855 it had more ~~bits~~ values —although admittedly a bit messy as it was somehow possible to be both `OPENSSL_EC_EXPLICIT` and `OPENSSL_EC_NAMED_CURVE`— and was used also to pick the default encodings for points among uncompressed, compressed and hybrid forms.
Before that we were using bitwise logic to handle the flag field, masking out bits of it in groups.
At this point it became a 1-bit field, and later on with 86f300d38540ead85543aee0cb30c32145931744, we started replacing `0` with a more descriptive `OPENSSL_EC_EXPLICIT_CURVE`.
Code expecting that field to be exactly 1-bit is now being out for so long that I am not sure we can do much about it.

Ideally we could have had this discussion much sooner, a little bit over 1 year ago when the issue was first opened, so we could have included `EC_GROUP_get_asn1_flag()` and `EC_GROUP_set_asn1_flag()` in the deprecation list for 3.0.0, while providing a new better named function as an alternative.
I don't think this is something we can do this close to beta, unfortunately, so we will likely have to stuck with it until the next major release :(

**Comment by romen** [comment]
Apart from the ongoing discussion on consistency in handling `EC_GROUP_get_asn1_flag()` this PR is now stuck (as #12457 for 1.1.1 has been for the past year) on Windows-specific test failures.

I still have the same questions about these failures:

- are the failures just a peculiarity of our test rigging?
- are the `_fp` variants of our functions supposed to work in Windows?
  - if not, should the prototypes for these functions be masked by some `#ifdef` in the public API?
  - if yes, can we get our testing framework to play nice with whatever needs to happen for these functions to work?
    - if we cannot fix the testing framework, should we disable this specific test for Windows? (what kind of `#ifdef` are rocmmended for this?)
    
I don't like the idea of disabling the test in Windows: if those functions are supposed to work in a Windows environment.
It feels like cheating, and the lack of testing of these functions is what made #12443 possible in the first place.

**Comment by romen** [comment]
@levitte can you check if d1202e8 is OK or if I need to tweak it differently?
The original implementation for 1.1.1 was by @bernd-edlinger but in `master` things changed quite a bit so I had to get creative in my copypaste from `apps/build.info`.

One thought I had ( which maybe won't work because this whole windows workaround remains a misery to me) was to instead incorporate this in the test static lib rather than each individual binary that might need it.

File: crypto/ec/ec_asn1.c (hunk 1 above)
Comment (romen) [reply crypto/ec/ec_asn1.c:488]: @paulidale how should I address the comment? do you still prefer bitwise access? If instead we go for `==` I'd rather do the consistency improvement of https://github.com/openssl/openssl/pull/16355#discussion_r691689240 as a separate PR.

**Comment by romen** [comment]
The last two commits do not change any code, and are just there as a remainder about fixing the typos in the commit messages when merging.

File: crypto/ec/ec_asn1.c (hunk 1 above)
Comment (paulidale) [reply crypto/ec/ec_asn1.c:488]: Go for `==` and add the consistency separately.

**Comment by romen** [comment]
> @levitte can you check if [d1202e8](https://github.com/openssl/openssl/commit/d1202e8e18517f39f765e4ae6da1b6a717631a29) is OK or if I need to tweak it differently?
> The original implementation for 1.1.1 was by @bernd-edlinger but in `master` things changed quite a bit so I had to get creative in my copypaste from `apps/build.info`.
> 
> One thought I had ( which maybe won't work because this whole windows workaround remains a misery to me) was to instead incorporate this in the test static lib rather than each individual binary that might need it.

Ping @levitte

**Comment by mattcaswell** [comment]
OTC: This is fine for inclusion in 3.0 subject to the normal review process.

File: crypto/ec/ec_curve.c [hunk 2]
```diff
@@ -3223,6 +3223,43 @@ static EC_GROUP *ec_group_new_from_data(OSSL_LIB_CTX *libctx,
...
+         * `EC_GROUP_new()`, when we don't have enough elements to determine if
+         * an OID for the curve name actually exists.
+         * We could implement this check on `EC_GROUP_set_curve_name()` but
+         * overloading the simple setter with this lookup could have a negative
+         * performance impact and unexpected consequences.
+         */
+        ASN1_OBJECT *asn1obj = OBJ_nid2obj(curve.nid);
+
+        if (asn1obj == NULL) {
+            ECerr(EC_F_EC_GROUP_NEW_FROM_DATA, ERR_R_OBJ_LIB);
```
Comment (t8m) [inline crypto/ec/ec_curve.c:3244]: You need to use ERR_raise() on master.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by t8m** [comment]
@romen will you merge this or should I?

**Comment by romen** [comment]
I was planning to do it in the weekend but I had to deal with unexpected things. I will merge it shortly as I had to squash and fix commit messages!

**Comment by romen** [comment]
Merged to `master` with:

- ea1128e94e `Add tests for i2d_TYPE_fp and d2i_TYPE_fp`
- cca8a4ceda `Fix d2i_ECPKParameters_fp and i2d_ECPKParameters_fp macros`
- 7aa3dfc421 `[ec] Do not default to OPENSSL_EC_NAMED_CURVE for curves without OID`
- 78082769fa `Use applink to fix windows tests`

Thanks @t8m for the review, @bernd-edlinger to solve the Applink issue!

**Pull request closed without merging**